torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
         self.last_keep_alives = {}
 
 
+class _RendezvousStateHolder(ABC):
+    """Holds the rendezvous state synced with other nodes."""
+
+    state: _RendezvousState
+
+    def __init__(self) -> None:
+        self.state = _RendezvousState()
+
+    @abstractmethod
+    def sync(self) -> None:
+        """Gets or sets the latest rendezvous state."""
+
+    @abstractmethod
+    def mark_dirty(self) -> None:
+        """Marks the local rendezvous state as dirty."""
+
+
+class _RendezvousStateHolderImpl(_RendezvousStateHolder):
+    """Represents the default implementation of ``RendezvousStateHolder``.
+
+    Args:
+        backend:
+            The rendezvous backend to use.
+        settings:
+            The rendezvous settings.
+        cache_duration:
+            The amount of time, in seconds, to cache the last rendezvous state
+            before requesting it from the backend again.
+    """
+
+    backend: RendezvousBackend
+    settings: RendezvousSettings
+    cache_duration: int
+
+    _token: Token
+    _dirty: bool
+    _last_sync_time: float
+
+    def __init__(
+        self, backend: RendezvousBackend, settings: RendezvousSettings, cache_duration: int = 1
+    ) -> None:
+        super().__init__()
+
+        self.backend = backend
+        self.settings = settings
+        self.cache_duration = cache_duration
+
+        self._token = None
+        self._dirty = False
+        self._last_sync_time = -1
+
+    def sync(self) -> None:
+        """See base class."""
+        if self._dirty:
+            state_bits = pickle.dumps(self.state)
+
+            response = self.backend.set_state(state_bits, self._token)
+        else:
+            if self.cache_duration > 0:
```

Comment:
I don't think we need cache, it would just complicate the rendezvous.

How is it going to be used? - is it going to be periodically invoked in a daemon thread?

Reply:
Caching (by default 1 second) actually helps quite a bit if you have separate components in your system that all use the same `_RendezvousStateHolder` instance to retrieve state data (as in `DynamicRendezvousHandler). As they are decoupled from each other, if they frequently query the state, this reduces the amount of traffic.